sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp

```diff
@@ -2059,25 +2078,124 @@ namespace Azure { namespace Storage { namespace Blobs {
       }
 
-      struct ListBlobsOptions
+      struct ListBlobsFlatOptions
+      {
+        Azure::Core::Nullable<int32_t> Timeout;
+        Azure::Core::Nullable<std::string> Prefix;
+        Azure::Core::Nullable<std::string> Marker;
+        Azure::Core::Nullable<int32_t> MaxResults;
+        ListBlobsIncludeItem Include = ListBlobsIncludeItem::None;
+      }; // struct ListBlobsFlatOptions
+
+      static Azure::Core::Http::Request ListBlobsFlatConstructRequest(
+          const std::string& url,
+          BodyStreamPointer& body,
+          const ListBlobsFlatOptions& options)
+      {
+        unused(body);
+        auto request = Azure::Core::Http::Request(Azure::Core::Http::HttpMethod::Get, url);
+        request.AddHeader("Content-Length", "0");
+        request.AddHeader("x-ms-version", "2019-12-12");
+        if (options.Timeout.HasValue())
+        {
+          request.AddQueryParameter("timeout", std::to_string(options.Timeout.GetValue()));
+        }
+        request.AddQueryParameter("restype", "container");
+        request.AddQueryParameter("comp", "list");
+        if (options.Prefix.HasValue())
+        {
+          request.AddQueryParameter("prefix", options.Prefix.GetValue());
+        }
+        if (options.Marker.HasValue())
+        {
+          request.AddQueryParameter("marker", options.Marker.GetValue());
+        }
+        if (options.MaxResults.HasValue())
+        {
+          request.AddQueryParameter("maxresults", std::to_string(options.MaxResults.GetValue()));
+        }
+        std::string list_blobs_include_item = ListBlobsIncludeItemToString(options.Include);
+        if (!list_blobs_include_item.empty())
+        {
+          request.AddQueryParameter("include", list_blobs_include_item);
+        }
+        return request;
+      }
+
+      static BlobsFlatSegment ListBlobsFlatParseResponse(
+          Azure::Core::Context context,
+          std::unique_ptr<Azure::Core::Http::Response> pHttpResponse)
+      {
+        unused(context);
+        Azure::Core::Http::Response& httpResponse = *pHttpResponse;
+        BlobsFlatSegment response;
+        auto http_status_code
+            = static_cast<std::underlying_type<Azure::Core::Http::HttpStatusCode>::type>(
+                httpResponse.GetStatusCode());
+        if (!(http_status_code == 200))
+        {
+          throw StorageError::CreateFromResponse(std::move(pHttpResponse));
+        }
+        {
+          auto bodyStream = httpResponse.GetBodyStream();
+          std::vector<uint8_t> bodyContent;
+          if (bodyStream->Length() == -1)
+          {
+            bodyContent = Azure::Core::Http::BodyStream::ReadToEnd(context, *bodyStream);
+          }
+          else
+          {
+            bodyContent.resize(static_cast<std::size_t>(bodyStream->Length()));
+            Azure::Core::Http::BodyStream::ReadToCount(
+                context, *bodyStream, &bodyContent[0], bodyStream->Length());
+          }
```

Comment:
Ideally with `if (bodyStream->Length() < 0)`.
Why not just `ReadToEnd`?

Reply:
If we know the body size beforehand, we can allocate the buffer in one shot.

Reply:
We don't need to expand the buffer again and again

Reply:
Sure, but I don't know if these extra lines of code are worthwhile due to we don't have the perf counter yet.
I'm not saying we shouldn't do it, I'm saying that we should have a same standard on optimization in code level by the look of it, not the actual number.

Reply:
Yes. In `ListBlobs` API, the result can be up to several megabytes. So I think it's necessary to optimize the code to avoid copying several mb of data.

Reply:
> I don't think it's necessary to benchmark before making every optimization. If we can foresee that the overhead is not trivial, we can directly optimize it, right?

And I think it's always a good habit to write optimized code.

Anyway, then I think it is only the issue with the if statement. I don't like `if (bodyStream->Length() == -1)` because it has implications on the implementation outside our control. I like my proposal better. Not critical so I'll sign-off the PR first.

Reply:
> And I think it's always a good habit to write optimized code.

No, I cannot agree on this. It's all about tradeoff between optimization and readability/maintainability. Writing assembly is the most optimized but we hardly ever do that because it's not readable.

Reply:
Your proposal is `if (bodyStream->Length() < 0)`. What's the difference here? Returning `-1` on unknown stream length is defined behavior. It's not going to return other value for now.

Let's say in the future it'll return other negative values, in which way, one is going to be better than the other?